### PR TITLE
[INFRA-1005] Allow contributors to deploy -SNAPSHOTs

### PR DESCRIPTION
**Pull request by stephenc**
See [INFRA-1005](https://issues.jenkins-ci.org/browse/INFRA-1005)

- Also tidy up the updater and allow for dry runs and off-line testing

- Also adds "stephenconnolly" as a contributor to [git](https://github.com/jenkinsci/git-plugin) and [mercurial](https://github.com/jenkinsci/mercurial-plugin) plugins @jglick @MarkEWaite

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -30,11 +30,7 @@ node('java') {
         stage 'Run'
         withCredentials([[$class: 'UsernamePasswordMultiBinding', credentialsId: 'artifactoryAdmin',
                           usernameVariable: 'ARTIFACTORY_USERNAME', passwordVariable: 'ARTIFACTORY_PASSWORD']]) {
-            sh '${JAVA_HOME}/bin/java' +
-                    ' -DdefinitionsDir=$PWD/permissions' +
-                    ' -DartifactoryApiTempDir=$PWD/json' +
-                    ' -Djava.util.logging.SimpleFormatter.format="%1$tY-%1$tm-%1$td %1$tH:%1$tM:%1$tS %4$s: %5$s%6$s%n"' +
-                    ' -jar target/repository-permissions-updater-*-bin/repository-permissions-updater-*.jar'
+            sh 'target/appassembler/bin/repository-permissions-updater -d $PWD/permissions -w $PWD/json'
```
Comment (daniel-beck) [inline Jenkinsfile]: I strongly prefer long options for readability.

File: README.md [hunk 2]
```diff
@@ -52,6 +52,10 @@ Create a new YAML file similar to existing files.
 
 Edit the `developers` list in the YAML file for the plugin.
 
+### Adding a `-SNAPSHOT` only uploader to an existing plugin
+
+Create or edit the `contributors` list in the YAML file for the plugin.
```
Comment (daniel-beck) [inline README.md]: A lot more of the readme needs to be updated to incorporate this properly.

File: src/main/groovy/io/jenkins/infra/repository_permissions_updater/ArtifactoryPermissionsUpdater.groovy [hunk 3]
```diff
@@ -288,9 +323,58 @@ public class ArtifactoryPermissionsUpdater {
...
+                            System.getenv("ARTIFACTORY_PASSWORD").toCharArray());
+                }
+            });
+        }
+
+        File definitionsDir = new File(options.d?:defDefnDir)
+        File workDir = new File(options.w?:defWorkDir)
+        generateApiPayloads(definitionsDir, workDir)
+        submitPermissionTargets(workDir, dryRun)
+        if (!offline) {
```
Comment (daniel-beck) [inline src/main/groovy/io/jenkins/infra/repository_permissions_updater/ArtifactoryPermissionsUpdater.groovy]: Why this condition if offline implied dry-run?

**Comment by daniel-beck** [comment]
@stephenc I wonder whether snapshot permission could be granted to anyone. Since snapshots are not distributed further, and referenced downstream using their exact date, even malicious uploads wouldn't be a problem AFAICT. WDYT?

**Comment by stephenc** [comment]
@daniel-beck snapshot deployment for all would be super-awesome... I'm just trying to unblock my small issue with least change

File: src/main/groovy/io/jenkins/infra/repository_permissions_updater/ArtifactoryPermissionsUpdater.groovy (hunk 3 above)
Comment (stephenc) [inline src/main/groovy/io/jenkins/infra/repository_permissions_updater/ArtifactoryPermissionsUpdater.groovy]: dry run when on-line will tell you the permissions for removal.

the submit doesn't care about on-line once dryRun is enabled, but the clean up needs auth

**Comment by MarkEWaite** [comment]
+1 for @stephenc being added as a contributor to the git plugin.  Should he also be added as a contributor to the git client plugin during this same change.

**Comment by daniel-beck** [comment]
@stephenc I granted unrestricted upload permissions to `snapshots` to you, temporarily.

Longer term, I think I can open up snapshot upload permissions in general, as I doubt there's a security problem with that. Will need to think about it first though.

**Comment by batmat** [comment]
I tend to agree this should not hurt, and would test some things indeed. Or alternatively we could have PR build going up to deploy.

**Comment by daniel-beck** [comment]
As written on the dev list, I opened up snapshot deployment to all users.

I may still use some of the other changes in this PR when setting up real PR builds (which unfortunately would require Artifactory admin permissions right now to find existing users accounts).

**Pull request closed without merging**